### PR TITLE
Feature/alert list

### PR DESCRIPTION
**Pull request by aaronchongth**
## What's new

Featuring a notifications button on the appbar (think linkedin or facebook notifications), which allows users to view unacknowledged alerts, open them again, and acknowledge them.

* appbar notifications button, showing number of unacknowledged alerts
* refactor or alert store and task alert handler -> task alert dialog
* added `TaskInspector` on the alert dialog
* removal of robot alerts (not required)

https://github.com/open-rmf/rmf-web/assets/5383623/e4c2da9c-4966-477c-a80e-0866e5d4872a

When there are no unacknowledged alerts, this is shown,
image

## Self-checks

- [ ] I have prototyped this new feature (if necessary) on Figma 
- [ ] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

File: packages/dashboard/src/components/tasks/task-alert.tsx [hunk 1]
```diff
@@ -143,112 +145,112 @@ export function TaskAlertHandler({ alerts, removeAlert }: TaskAlertHandlerProps)
...
+          key={taskAlert.task_id}
+          onDismiss={removeAlert}
+          onAcknowledge={acknowledgeAlert}
+          onInspect={() => setOpenTaskInspector(true)}
+          title={taskAlert.title}
+          progress={taskAlert.progress}
+          alertContents={taskAlert.content}
+          backgroundColor={taskAlert.color}
+        />
+      )}
```
Comment (Angatupyry) [inline packages/dashboard/src/components/tasks/task-alert.tsx:250]: Maybe we can simplify it
```suggestion
      <AlertDialog
        key={taskAlert.task_id}
        onDismiss={removeAlert}
        acknowledgedBy={taskAlert.acknowledgedBy}
        onAcknowledge={taskAlert.acknowledgedBy ? undefined : acknowledgeAlert}
        onInspect={() => setOpenTaskInspector(true)}
        title={taskAlert.title}
        progress={taskAlert.progress}
        alertContents={taskAlert.content}
        backgroundColor={taskAlert.color}
      />
```

**Review by Angatupyry — CHANGES_REQUESTED** [review]
It looks wonderful!
Thanks for doing it so thoroughly, Aaron!
I have left some suggestions that may be useful. 

I noticed that the progress bar is not being painted correctly in AlertDialog (react-componets), it is because the value passed in line 57 is not being multiplied by 100, so it is taking a very small value. 
One way to fix this is to remove the multiplication by 100 on line 61 and do it in the component call on line 126 value={progress * 100}.
Screenshot from 2023-05-23 11-06-29

With suggested changes above:
Screenshot from 2023-05-23 11-10-38

I also noticed that we are not using the try catch in some await async functions. I raise it as a warning and I think we should go ahead and add it for future implementations to handle errors correctly.

File: packages/dashboard/src/components/alert-store.tsx [hunk 2]
```diff
@@ -31,42 +17,71 @@ enum AlertCategory {
...
-      }
+      categorizeAndPushAlerts(alert);
+      refreshAlertCount.current += 1;
+      AppEvents.refreshAlertCount.next(refreshAlertCount.current);
     });
     return () => sub.unsubscribe();
   }, [rmf]);
 
-  const taskAlertKey: Partial<typeof taskAlerts> = { ...taskAlerts };
   const removeTaskAlert = (id: string) => {
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/alert-store.tsx:67]: although we don't expect to handle many task alerts at the same time, I would prefer to use a traditional `for ... in` loop here for performance, since there will be 1 less lookup for deletion, and it looks like a spread operator is generally less performant, https://www.measurethat.net/Benchmarks/Show/16078/0/javascript-spread-operator-vs-objectassign-vs-for-in-lo#latest_results_block

File: packages/dashboard/src/components/app-events.ts [hunk 3]
```diff
@@ -9,4 +16,6 @@ export const AppEvents = {
   robotSelect: new Subject<[fleetName: string, robotName: string] | null>(),
   taskSelect: new Subject<TaskState | null>(),
   refreshTaskQueueTableCount: new Subject<number>(),
+  refreshAlertCount: new Subject<number>(),
+  alertListOpenedAlert: new Subject<Alert | null>(),
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/app-events.ts:20]: Yup it is intentional for readability, since a `list` could be many things (from robot list or task list, etc), the alert list would refer to the menu list explicitly

File: packages/dashboard/src/components/alert-store.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/alert-store.tsx:60]: I did a bit more digging and found that the general consensus of typescript prioritizes readability and async operations, I've refactored it to use `Object.entries` after seeing some of your examples, https://github.com/open-rmf/rmf-web/pull/710/commits/f843fb92cd35e06af5fe64585a298becbe6f02fd

File: packages/dashboard/src/components/alert-store.tsx [hunk 4]
```diff
@@ -31,42 +17,71 @@ enum AlertCategory {
...
+            }
+          }
+          filteredTaskAlerts[alert.id] = alert;
+          return filteredTaskAlerts;
+        });
+        break;
+      default:
+      // do nothing in default case
+    }
+  };
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/alert-store.tsx]: https://github.com/open-rmf/rmf-web/pull/710/commits/f843fb92cd35e06af5fe64585a298becbe6f02fd

File: packages/dashboard/src/components/appbar.tsx [hunk 5]
```diff
@@ -166,10 +174,36 @@ export const AppBar = React.memo(({ extraToolbarItems }: AppBarProps): React.Rea
...
-      setPlaceNames(getPlaces(map).map((p) => p.vertex.name)),
+
+    const subs: Subscription[] = [];
+    subs.push(
+      rmf.buildingMapObs.subscribe((map) =>
+        setPlaceNames(getPlaces(map).map((p) => p.vertex.name)),
+      ),
     );
-    return () => sub.unsubscribe();
+    subs.push(
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/appbar.tsx:183]: Thanks for catching it. https://github.com/open-rmf/rmf-web/pull/710/commits/711cd80b8893d0c483c324626f9a55b74c5f2e28 I just removed the region

File: packages/dashboard/src/components/appbar.tsx [hunk 6]
```diff
@@ -264,7 +298,30 @@ export const AppBar = React.memo(({ extraToolbarItems }: AppBarProps): React.Rea
...
+      for (let alert of alerts) {
+        if (alert.acknowledged_by || alert.unix_millis_acknowledged_time) {
+        } else {
+          unackList.push(alert);
+        }
+      }
+      setUnacknowledgedAlertList(unackList.reverse());
+    })();
+    setAlertListAnchor(event.currentTarget);
+  };
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/appbar.tsx:318]: thanks for catching the stale code, I didn't update it after removing all the acknowledged task alert stuff https://github.com/open-rmf/rmf-web/pull/710/commits/f843fb92cd35e06af5fe64585a298becbe6f02fd

File: packages/dashboard/src/components/appbar.tsx [hunk 7]
```diff
@@ -319,9 +376,68 @@ export const AppBar = React.memo(({ extraToolbarItems }: AppBarProps): React.Rea
...
+                  <Typography variant="body2" mx={2} noWrap>
+                    {new Date(alert.unix_millis_created_time).toLocaleString()}
+                  </Typography>
+                  <Typography variant="body2" noWrap>
+                    {alert.category.toUpperCase()} alert
+                  </Typography>
+                </MenuItem>
+              </Tooltip>
+            ))}
+          </Menu>
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/appbar.tsx]: https://github.com/open-rmf/rmf-web/pull/710/commits/560d6179659697013768556d9ea87371b85ac5d7

[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/appbar.tsx]: https://github.com/open-rmf/rmf-web/pull/710/commits/560d6179659697013768556d9ea87371b85ac5d7

File: packages/dashboard/src/components/tasks/task-alert.tsx [hunk 8]
```diff
@@ -143,112 +145,112 @@ export function TaskAlertHandler({ alerts, removeAlert }: TaskAlertHandlerProps)
...
+          const ackSecondsAgo =
+            (new Date().getTime() - ackResponse.unix_millis_acknowledged_time) / 1000;
+          showAlertMessage += ` ${Math.round(ackSecondsAgo)}s ago`;
+        }
+        showAlert('success', showAlertMessage);
+      } else {
+        console.log(`Failed to acknowledge alert ID ${taskAlert.task_id}`);
+        showAlert('error', `Failed to acknowledge alert ID ${taskAlert.task_id}`);
+      }
+    })();
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-alert.tsx]: https://github.com/open-rmf/rmf-web/pull/710/commits/32189f6221c4a6e5bae6edeae566cd6d5fee157d

File: packages/dashboard/src/components/tasks/task-alert.tsx (hunk 1 above)
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-alert.tsx]: Thanks for catching this, it definitely cleaned up lots, https://github.com/open-rmf/rmf-web/pull/710/commits/781f7a60c4817a44e6b9645fbbd0bb89b98ad2ab

**Comment by aaronchongth** [comment]
Since we are only dealing with tasks now, I've also revamped the appbar menu items, and gave the hover tooltips more information, https://github.com/open-rmf/rmf-web/pull/710/commits/042a12ea6676eb46350e5bfaba20deb92e3a3f1c

Screenshot from 2023-05-25 15-44-37

**Comment by Angatupyry** [comment]
@aaronchongth Thanks for taking the time on this!
The only minor detail is the first thing I mentioned about the progress line. It is not painting according to the progress of the task. 
I could not suggest a change from this PR since that line has not been modified but I left the minimal changes needed in the first general comment. 

Other than that the PR looks great. 
LGTM!

**Review by Angatupyry — APPROVED** [review]
LGTM!
